### ISSUE TITLE
ranking

### ISSUE THREAD

**Issue by gerardkm94** [gdsa-upc/gdsa-suport#23]
En la función ranking, debemos calcular la distancia euclidea del bag of words de una query image, con todos los bow de las de training, hasta ahí bien pero cuando utilizamos la función que lo realiza, comparamos el bow de la query con una lista de todos los bow de training.
sklearn.metrics.pairwise.pairwise_distances(bow,(LISTA) bow2, metric='euclidean', n_jobs=1, **kwds)

Si lo hacemos de esta manera, perdemos la información referente a las ID de las imagenes de training, y obtendriamos solamente una lista de las distancias pero no sabríamos a que ID's perteneces esas distancias.

Alguna ayuda?

**Comment by amaiasalvador** [comment]
Hola Gerard,

Ja casi ho teniu!

Entenc que per poder tenir la llista de descriptors BoW tots junts en una llista el que heu fet ha estat recorrer el diccionari de BoW d'entrenament, i per cada element dins de _dictionary_train.keys()_ heu guardat el seu valor (el descriptor BoW) en una llista. Això seria equivalent a directament utilitzar _dictionary_train.values()_, que ja retorna tots els BoW junts (així us estalvieu el bucle _for_). Sigui com sigui, al final d'aquest procediment tindreu una llista de 450 vectors de tamany K (tamany del codebook).

Llavors, cridant a la funció _pairwise_distances_  amb el vector BoWde la query i el llistat de BoWs d'entrenament, obteniu una llista de 450 valors (les distàncies del vector BoW de la query contra cadascun dels BoW de les imatges d'entrenament). 

Fins aquí crec que ja heu arribat. Ara us falta l'últim pas, que és el d'ordenar les ids de les imatges _dictionary_train.keys()_ en funció de les distàncies que heu obtingut. Penseu que la funció _pairwise_distances_ retorna les distàncies per ordre (és a dir, el primer valor de la llista és la distància del BoW de la query contra el primer BoW de la llista de BoWs d'entrenament). D'aquesta manera, la llista de distàncies segueix el mateix ordre que _dictionary_train.keys()_: la distància en la posició 0 de la llista correspon a la id en la posició 0 dins de _dictionary_train.keys()_.  Ara només us falta ordenar la llista _dictionary_train.keys()_ utilitzant la llista de distàncies. A la llista resultant, la id de la imatge amb mínima distància ha de ser la primera. 

Pista: la funció [numpy.argsort](http://docs.scipy.org/doc/numpy/reference/generated/numpy.argsort.html) retorna els elements que ordenen un array en ordre ascendent. Utilitzeu-la sobre la llista de distàncies per saber com s'ordenaria el vector de distàncies. Aquest ordre és el que heu d'aplicar per ordenar la llista de les ids de les imatges.
